pip/index.py

```diff
@@ -66,7 +66,19 @@ def __init__(self, find_links, index_urls,
                 "'session'"
             )
 
-        self.find_links = find_links
+        # Build find_links. If an argument starts with ~, it may be
+        # a local file relative to a home directory. So try normalizing
+        # it and if it exists, use the normalized version.
+        # This is deliberately conservative - it might be fine just to
+        # blindly normalize anything starting with a ~...
+        self.find_links = []
+        for link in find_links:
+            if link.startswith('~'):
```

Comment:
My instinct says no, ```~/foo``` is a path and should be subject to typical path expansions. But ```file://~/foo``` is a URL, and ~-expansion isn't part of the URL spec.